### ISSUE TITLE
[Question / Request] Tokens / multiple channels

### ISSUE THREAD

**Issue by timothyclarke** [CircleCI-Public/slack-orb#77]
I'm wanting to send slack messages to different teams in different channels. 

It looks like a web hook is linked to a specific channel. Is there any way to use a token and then specify the channel?

https://get.slack.help/hc/en-gb/articles/215770388-Create-and-regenerate-API-tokens

**Comment by KyleTryon** [comment]
Hello @timothyclarke At the moment we don't support multiple webhooks as a parameter but I would be willing to take that as a feature request. At the moment I believe the best way to do this would to be to call the command twice, each with the different webhook value.

**Comment by kamontat** [comment]
I will try this. Maybe we can change to use apps token instead? for multiple channels in slack?

**Comment by timothyclarke** [comment]
Just to give this some more context. I'm not talking about 2 channels. I'm talking closer to 10. 

As builds go into DEV environments the appropriate team needs to know. 
We're using micro services/ containers so as a build / container progress through the various environments to Production different teams should be aware. Eg QA

**Comment by gmemstr** [comment]
I wonder if [contexts](https://circleci.com/docs/2.0/contexts/) would help in this scenario.

**Comment by timothyclarke** [comment]
I think this one was solved with Orb version 3.4.0  (which introduced the `channel` param).  I'll test and if so will close this

**Comment by KyleTryon** [comment]
So, I was looking into this specifically for the new version of the slack orb (version 4.0.0) due to be released next week, however I did not include the channel parameter because of this: 
![image](https://user-images.githubusercontent.com/33272306/93779013-7dd62680-fbf4-11ea-9868-cc9177448498.png)
https://api.slack.com/messaging/webhooks

What I would recommend is creating separate webhooks for each channel and save the env vars separately in CircleCI (which could be done with contexts if desired) and when calling the orb, you can pass the channel to the webhook parameter.

```
slack/notify:
  webhook: $SLACK_CHANNEL_1
```

**Comment by timothyclarke** [comment]
@KyleTryon Thanks. One of the issues I'm trying to address is that multiple (currently 2) teams / channels want to be notified of Production releases.  
* Development teams manage their own builds / deployments and have a slack channel they want notified
* There is another channel which includes the service desk and various other managers who *need* to know

Running a running a step twice with different inputs isn't an issue, running multiple jobs just for notifications become a little tedious. 
Parsing multiple webhooks also feels wrong

**Comment by timothyclarke** [comment]
I've been able to use curl to send a message to a custom (public) channel
I've setup the App per step 1 of https://github.com/CircleCI-Public/slack-orb/tree/v3.4.2#how-to-get-your-slack-webhook
I think went to the oauth section `https://api.slack.com/apps/<app_id>/oauth?` and added the oauth scope `chat:write.public`

From there it was a simple `curl -X POST -H 'Authorization: Bearer xoxb-###-###-***'  -H 'Content-type: application/json' --data '{"channel": "#my_public_channel","text":"Hello, World!"}' https://slack.com/api/chat.postMessage`

**Comment by gmemstr** [comment]
Hey @timothyclarke, we've dug a bit deeper and we're working on implementing this in v4 -- thanks for the pointers!

**Comment by timothyclarke** [comment]
@gmemstr Thank you for the update / feedback. Would you like me to close https://github.com/CircleCI-Public/slack-orb/pull/170?

**Comment by gmemstr** [comment]
@timothyclarke Since it's a breaking change anyways, we're going to opt to roll this into #160 instead (#172 wip). Go ahead and close the PR, unless we can find a way around making it a breaking change so we can potentially roll a minor release for 3.x users.

**Comment by timothyclarke** [comment]
Thank you @gmemstr 
Please you let me know when this is/will be in `circleci/slack@dev:beta` so I can test the functionality.

I tried a few minutes before posting this and the beta version looks to be still based on the webhook rather than the Bearer token.

**Comment by gmemstr** [comment]
@timothyclarke Thank you for the reminder! I've updated the `dev:beta` version with the latest version of the orb.

**Comment by timothyclarke** [comment]
@gmemstr Thank you.
I've just tested and I'm getting branch does not match issue
```
Current branch does not match any item from the "branch_list" parameter
Current branch: JIRA-123
```
Note I've tweaked the branch name above, but kept the regex pattern
The appropriate part of my `.circleci/config.yml` is 
```
orbs:
  betaslack: circleci/slack@dev:beta

jobs:
  Notify Slack:
    docker:
      - image: alpine:3.12.0
    environment:
      SLACK_DEFAULT_CHANNEL: '#circle-test'
    steps:
      - betaslack/notify:
          event: pass
          template: success_tagged_deploy_1
```

**Comment by KyleTryon** [comment]
@timothyclarke thank you for the bug fix and a solution! You are wonderful people. We're taking a look now.

**Comment by timothyclarke** [comment]
:D glad that this is progressing. pls let me know when the updated beta is published so I can re-test.
I just switched from 
```
    environment:
      SLACK_DEFAULT_CHANNEL: '#circle-test'
```
to
```
    steps:
      - betaslack/notify:
          channel:  '#circle-test'
``` 
per the current code and received `Unexpected argument(s): channel` :(

**Comment by KyleTryon** [comment]
> :D glad that this is progressing. pls let me know when the updated beta is published so I can re-test.
> I just switched from
> 
> ```
>     environment:
>       SLACK_DEFAULT_CHANNEL: '#circle-test'
> ```
> 
> to
> 
> ```
>     steps:
>       - betaslack/notify:
>           channel:  '#circle-test'
> ```
> 
> per the current code and received `Unexpected argument(s): channel` :(

Hey @timothyclarke ,

Wanted your opinion as I am now trying to add support for alpine and make some POSIX changes. I realize on testing and research, it is common for alpine images to not include a root user or sudo privileges, which makes it impossible to install JQ on alpine most of the time.

We can probably still work on Alpine but with the requirement that CURL and JQ be installed in the image and we wouldnt attempt to install them for you on alpine. 

Any thoughts?

---

Also for the channel, and we will add docs, you would not include the hashtag. The best method is using just the "channel id" rather than the text name.

**Comment by timothyclarke** [comment]
@KyleTryon Ensure that both `CURL` and `JQ` are documented. I'd put the attempt to install behind a flag (and default it to false). 

I think this will fall into two camps, builds (or similar compilation jobs) and processes (Terraform, Ansible, Helm etc). I'd imagine that you should have a better idea on that point than me. 

We're very careful about how the containers are run in our DEV / TEST / PRODUCTION environments and in the process jobs. When we're building code in a docker container that will be disposed of at the end of the run, we're not as concerned. The only things I'm keeping from build to build are 
* The Target. eg a docker container or similar artifact produced as a direct product of the build.
* Any test results
* Cached workspaces, eg local maven repo's
I use to be very concerned about the build environment on other platforms when that build environment was re-used over and over. Once I switched to building in a docker container the focus switched from maintaining that tool chain at a known state, to building that tool chain to the known state.

**Comment by timothyclarke** [comment]
@KyleTryon @gmemstr Is there any chance the beta orb can be updated to the current code ?
I'm still getting the error described in https://github.com/CircleCI-Public/slack-orb/issues/77#issuecomment-701353116 and looking at the code int he job I'm still seeing `     if [[ "$CIRCLE_BRANCH" =~ ^${i}$ ]]; then`

**Comment by gmemstr** [comment]
@timothyclarke Updated :) Should really look into automating this, but we're approaching a finalised version regardless.

**Comment by timothyclarke** [comment]
I've been able to 
* Send messages :D
* Send messages to different channels (used the `channel` param). Note as most of the teams I work with use of the slack clients getting the channel (`[A-Z]+` rather than `#channel`) involved going beak to the browser. from within the app if I right clicked the channel name there is a "Copy link" option. The channel is the string after the final slash
* Customize messages with the block kit

I've been able to use multiple contexts as well (which is one of the features I've been waiting a while for).
Thanks for all the effort that has been put into this.
Now I've just got to integrate this into the workflows of 5 teams
